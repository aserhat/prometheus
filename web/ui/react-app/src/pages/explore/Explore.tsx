import React, { ChangeEvent, FC, useState } from 'react';
import { Input, InputGroup, Table } from 'reactstrap';
import { withStatusIndicator } from '../../components/withStatusIndicator';
import { useFetch } from '../../hooks/useFetch';
import { usePathPrefix } from '../../contexts/PathPrefixContext';
import { API_PATH } from '../../constants/constants';
import { faSort, faSortDown, faSortUp } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { IconDefinition } from '@fortawesome/fontawesome-common-types';
import sanitizeHTML from 'sanitize-html';
import { Fuzzy, FuzzyResult } from '@nexucis/fuzzy';

const fuz = new Fuzzy({ pre: '<strong>', post: '</strong>', shouldSort: true });
const metricSeparator = '||';

interface MetricData {
  type?: string;
  help?: string;
  unit?: string;
}

interface ExploreMap {
  [key: string]: Array<MetricData>;
}

interface ExploreProps {
  status?: string;
  data?: ExploreMap;
}

const compareAlphaFn =
  (keys: boolean, reverse: boolean) =>
  ([k1, v1]: [string, Array<MetricData>], [k2, v2]: [string, Array<MetricData>]): number => {
    const a = keys ? k1 : k1;
    const b = keys ? k2 : k2;
    const reverser = reverse ? -1 : 1;
    return reverser * a.localeCompare(b);
  };

const getSortIcon = (b: boolean | undefined): IconDefinition => {
  if (b === undefined) {
    return faSort;
  }
  if (b) {
    return faSortDown;
  }
  return faSortUp;
};

interface SortState {
  name: string;
  alpha: boolean;
  focused: boolean;
}

export const ExploreContent: FC<ExploreProps> = ({ data = {} }) => {
  const initialSearch = '';
  const [searchState, setSearchState] = useState(initialSearch);
  const initialSort: SortState = {
    name: 'Metric',
    alpha: true,
    focused: true,
  };
  const [sortState, setSortState] = useState(initialSort);
  const searchable = Object.entries(data)
    .sort(compareAlphaFn(sortState.name === 'Metric', !sortState.alpha))
    .map(([metric, value]) => `${metric}${metricSeparator}${value[0]['help']}`);
  let filtered = searchable;
  if (searchState.length > 0) {
    filtered = fuz.filter(searchState, searchable).map((value: FuzzyResult) => value.rendered);
  }
  return (
    <>
      <h2>Metrics Explorer</h2>
      <InputGroup>
        <Input
          autoFocus
          placeholder="Filter by metric name..."
          className="my-3"
          value={searchState}
          onChange={({ target }: ChangeEvent<HTMLInputElement>): void => {
            setSearchState(target.value);
          }}
        />
      </InputGroup>
      <Table bordered size="sm" striped hover>
        <thead>
          <tr>
            <td
              key="Metric"
              className={`px-4 Metric`}
              style={{ width: '30%' }}
              onClick={(): void =>
                setSortState({
                  name: 'Metric',
                  focused: true,
                  alpha: sortState.name === 'Metric' ? !sortState.alpha : true,
                })
              }
            >
              <span className="mr-2">Metric</span>
              <FontAwesomeIcon icon={getSortIcon(sortState.name !== 'Metric' ? undefined : sortState.alpha)} />
            </td>
            <td key="Help" className={'px-4 "Help"'} style={{ width: '70%' }}>
              <span className="mr-2">Help</span>
            </td>
          </tr>
        </thead>
        <tbody>
          {filtered.map((result: string) => {
            const [metricMatchStr, valueMatchStr] = result.split(metricSeparator);
            const sanitizeOpts = { allowedTags: ['strong'] };
            return (
              <tr key={metricMatchStr}>
                <td className="flag-item">
                  <span dangerouslySetInnerHTML={{ __html: sanitizeHTML(metricMatchStr, sanitizeOpts) }} />
                </td>
                <td className="flag-value">
                  <span dangerouslySetInnerHTML={{ __html: sanitizeHTML(valueMatchStr, sanitizeOpts) }} />
                </td>
              </tr>
            );
          })}
        </tbody>
      </Table>
    </>
  );
};
const ExploreWithStatusIndicator = withStatusIndicator(ExploreContent);

ExploreContent.displayName = 'Explore';

const Explore: FC = () => {
  const pathPrefix = usePathPrefix();
  const { response, error, isLoading } = useFetch<ExploreMap>(`${pathPrefix}/${API_PATH}/metadata`);
  return <ExploreWithStatusIndicator data={response.data} error={error} isLoading={isLoading} />;
};

export default Explore;
