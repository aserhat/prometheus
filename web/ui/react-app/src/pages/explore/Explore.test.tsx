import * as React from 'react';
import { shallow } from 'enzyme';
import { ExploreContent } from './Explore';
import { Input, Table } from 'reactstrap';
import toJson from 'enzyme-to-json';

const sampleMetadataResponse = {
    'prometheus_web_federation_warnings_total': [
      {
        'type': 'counter',
        'help': 'Total number of warnings that occurred while sending federation responses.',
        'unit': '',
      }
    ],
    'promttp_metric_handler_requests_in_flight': [
      {
        'type': 'gauge',
        'help': 'Current numver of scrapes being served.',
        'unit': '',
      }
    ]
};

describe('Explore', () => {
  it('renders a table with properly configured props', () => {
    const w = shallow(<ExploreContent data={sampleMetadataResponse} />);
    const table = w.find(Table);
    expect(table.props()).toMatchObject({
      bordered: true,
      size: 'sm',
      striped: true,
    });
  });

  it('should not fail if data is missing', () => {
    expect(shallow(<ExploreContent />)).toHaveLength(1);
  });

  it('should match snapshot', () => {
    const w = shallow(<ExploreContent data={sampleMetadataResponse} />);
    expect(toJson(w)).toMatchSnapshot();
  });

  it('is sorted by metric by default', (): void => {
    const w = shallow(<ExploreContent data={sampleMetadataResponse} />);
    const td = w.find('tbody').find('td').find('span').first();
    expect(td.html()).toBe('<span>prometheus_web_federation_warnings_total</span>');
  });

  it('sorts', (): void => {
    const w = shallow(<ExploreContent data={sampleMetadataResponse} />);
    const th = w
      .find('thead')
      .find('td')
      .filterWhere((td): boolean => td.hasClass('Metric'));
    th.simulate('click');
    const td = w.find('tbody').find('td').find('span').first();
    expect(td.html()).toBe('<span>promttp_metric_handler_requests_in_flight</span>');
  });

  it('filters by metric name', (): void => {
    const w = shallow(<ExploreContent data={sampleMetadataResponse} />);
    const input = w.find(Input);
    input.simulate('change', { target: { value: 'timeout' } });
    const tds = w
      .find('tbody')
      .find('td')
      .filterWhere((code) => code.hasClass('flag-item'));
    expect(tds.length).toEqual(1);
  });

});
