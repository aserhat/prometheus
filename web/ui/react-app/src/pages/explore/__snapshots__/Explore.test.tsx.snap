// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explore should match snapshot 1`] = `
<Fragment>
  <h2>
    Metrics Explorer
  </h2>
  <InputGroup
    tag="div"
  >
    <Input
      autoFocus={true}
      className="my-3"
      onChange={[Function]}
      placeholder="Filter by metric name..."
      type="text"
      value=""
    />
  </InputGroup>
  <Table
    bordered={true}
    hover={true}
    responsiveTag="div"
    size="sm"
    striped={true}
    tag="table"
  >
    <thead>
      <tr>
        <td
          className="px-4 Metric"
          key="Metric"
          onClick={[Function]}
          style={
            Object {
              "width": "30%",
            }
          }
        >
          <span
            className="mr-2"
          >
            Metric
          </span>
          <FontAwesomeIcon
            beat={false}
            beatFade={false}
            border={false}
            bounce={false}
            className=""
            fade={false}
            fixedWidth={false}
            flip={false}
            icon={
              Object {
                "icon": Array [
                  320,
                  512,
                  Array [
                    "sort-desc",
                  ],
                  "f0dd",
                  "M182.6 470.6c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-9.2-9.2-11.9-22.9-6.9-34.9s16.6-19.8 29.6-19.8H288c12.9 0 24.6 7.8 29.6 19.8s2.2 25.7-6.9 34.9l-128 128z",
                ],
                "iconName": "sort-down",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            maskId={null}
            pull={null}
            pulse={false}
            rotation={null}
            shake={false}
            size={null}
            spin={false}
            spinPulse={false}
            spinReverse={false}
            swapOpacity={false}
            symbol={false}
            title=""
            titleId={null}
            transform={null}
          />
        </td>
        <td
          className="px-4 \\"Help\\""
          key="Help"
          style={
            Object {
              "width": "70%",
            }
          }
        >
          <span
            className="mr-2"
          >
            Help
          </span>
        </td>
      </tr>
    </thead>
    <tbody>
      <tr
        key="prometheus_web_federation_warnings_total"
      >
        <td
          className="flag-item"
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "prometheus_web_federation_warnings_total",
              }
            }
          />
        </td>
        <td
          className="flag-value"
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Total number of warnings that occurred while sending federation responses.",
              }
            }
          />
        </td>
      </tr>
      <tr
        key="promttp_metric_handler_requests_in_flight"
      >
        <td
          className="flag-item"
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "promttp_metric_handler_requests_in_flight",
              }
            }
          />
        </td>
        <td
          className="flag-value"
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Current numver of scrapes being served.",
              }
            }
          />
        </td>
      </tr>
    </tbody>
  </Table>
</Fragment>
`;
